### DISCUSSION TITLE
[Apex] DML should be done in bulk

### DISCUSSION THREAD

**Discussion by PJ17101** [pmd/pmd#3021] [Q&A]
**Description:**
We should always bulkify the code which collects all the records to be updated/inserted/deleted and it does once DML for all qualified records. 
We shouldn't be making multiple same DMLs on the same object in the same contexts. Please check for existing DMLs if it is existing and can be reused.
Example :

```java
public class Foo {
    public void bar() {
	Account a=[select id from Account];
	Contact b=[select id from Account where id!=null];
	List<Account> accList=[a];
	Account c=[select id from Account Limit 5];
	Contact d=[select id from Account];
	update b;
	insert c;
	insert accList;
    }
}
```

**In the above code, Is there a way to identify the type DML operation variable?**

Like: 1 Type of accList is LIST
         2 Type of a,b,c is Object** 

if we are able to find the type of DML variable then we can give violations for all the DML operations whose variable not having type as LIst.

**Comment by adangel** [comment]
See https://sourceforge.net/p/pmd/discussion/188192/thread/c3850761bb/?limit=25#5485 for the original post.
